src/emit.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::rc::Rc;
+
+use anyhow::Result;
+use base64::Engine;
+use swc_common::comments::SingleThreadedComments;
+use swc_common::SourceMap;
+
+use crate::swc::ast::Program;
+use crate::swc::codegen::text_writer::JsWriter;
+use crate::swc::codegen::Node;
+use crate::swc::common::FileName;
+use crate::ModuleSpecifier;
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SourceMapOption {
+  /// Source map should be inlined into the source (default)
+  #[default]
+  Inline,
+  /// Source map should be generated as a separate file.
+  Separate,
+  /// Source map should not be generated at all.
+  None,
+}
+
+#[derive(Debug, Clone, Copy, Hash)]
+pub struct EmitOptions {
+  /// How and if source maps should be generated.
+  pub source_map: SourceMapOption,
+  /// Whether to inline the source contents in the source map. Defaults to `true`.
+  pub inline_sources: bool,
+  /// Whether to keep comments in the output. Defaults to `false`.
+  pub keep_comments: bool,
+}
+
+impl Default for EmitOptions {
+  fn default() -> Self {
+    EmitOptions {
+      source_map: SourceMapOption::default(),
+      inline_sources: true,
+      keep_comments: false,
+    }
+  }
+}
+
+/// An Emitter is used to emit a source file based on the emit options.
+pub struct Emitter {
+  emit_options: EmitOptions,
+  pub source_map: Rc<SourceMap>,
+}
+
+/// Source emitted based on the emit options.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+pub struct EmittedSource {
+  /// Emitted text.
+  pub text: String,
+  /// Source map back to the original file.
+  pub source_map: Option<String>,
+}
+
+impl Emitter {
+  pub fn new(specifier: &str, source: String, options: EmitOptions) -> Self {
```

Comment:
It's only used to construct the source map, and for NPM transpile I need the ability to specify a relative path as the source map path. Hence why this is a string instead of a module specifier